wd_helper.au3

```diff
@@ -875,49 +874,60 @@ Func __WD_FrameList_Internal($sSession, $sLevel, $sFrameAttributes, $bIsHidden,
 				$iErr = @error
 				If @error Then
 					$sMessage = 'Error occurred on "' & $sLevel & '" level when checking URL'
-				Else
-					$vResult = $sLevel & '|' & $sLevel & '|' & $sFrameAttributes & '|' & $s_URL & '|' & $sCurrentBody_ElementID & '|' & $bIsHidden & @CRLF
 				EndIf
 			EndIf
 		EndIf
 	EndIf
+	$vResult = $sLevel & '|' & $sLevel & '|' & $sFrameAttributes & '|' & $s_URL & '|' & $sCurrentBody_ElementID & '|' & $bIsHidden & '|' & @CRLF
+	ConsoleWrite("> " & $vResult & @CRLF) ; for testing only should be removed before merge.
```

Comment:
all lines with `; for testing only should be removed before merge.` comment should be removed before merge.

Reply:
Waiting for yor review...

Reply:
Need to update `$sArrayHeader` in wd_demo

Reply:
Also seeing errors generated from line 893 that look like this --

```
    __WD_Post: URL=HTTP://127.0.0.1:4444/session/7ec4e254-b880-4e63-a506-c3996c51ccd4/execute/sync; Data={"script":"return document.querySelectorAll('iframe')[0].outerHTML;", "args":[]}
    __WD_Post ==> Javascript Exception [23] : HTTP status = 500 ResponseText={"value":{"error":"javascript error","message":"TypeError: document.querySelectorAll(...)[0] is undefined","stacktrace":"@https://www.tutorialspoint.com/html/src/vertical_distribution_of_frames.htm:2:23\n@https://www.tutorialspoint.com/html/src/vertical_distribution_of_frames.htm:3:8\n"}}
    _WD_ExecuteScript ==> Javascript Exception [23] : Error occurred when trying to ExecuteScript
    _WD_DebugSwitch:  /  stack size: 1
```

Reply:
`$sMessage` contents gets overwritten by subsequent passes through For loop.

Reply:
> Need to update `$sArrayHeader` in wd_demo

Have it done already in my _WD_Locate testing script so I just add this header here ASAP .

Reply:
> > Need to update `$sArrayHeader` in wd_demo
> 
> Have it done already in my _WD_Locate testing script so I just add this header here ASAP .

It still needs to be updated in wd_demo, correct?

Reply:
Yes it was waiting to be updated.
And ... no as I just update this here: [$sArrayHeader suplemented](https://github.com/Danp2/au3WebDriver/pull/434/commits/b7e6b09efcecf03bdb6fc881cbad57e01f9628c0)

Reply:
> `$sMessage` contents gets overwritten by subsequent passes through For loop.

You mean in all the places where `ContinueLoop` is used just after `$sMessage = ` is set ?

Reply:
I will check what happen when I put message to console like this:
```autoit
$sMessage = 'Error occurred on "' & $sLevel & '" level when trying to check attributes child frames #' & $iFrame
__WD_ConsoleWrite($sFuncName & ": " & $sMessage, $_WD_DEBUG_Info)
ContinueLoop
```

Reply:
> > `$sMessage` contents gets overwritten by subsequent passes through For loop.
> 
> You mean in all the places where `ContinueLoop` is used just after `$sMessage = ` is set ?

Yes

Reply:
Please take a look on this: [$sMessage > __WD_ConsoleWrite](https://github.com/Danp2/au3WebDriver/pull/434/commits/70d168bcc5ef95b45262c2db198a14075bece9c8)

Reply:
> TypeError: document.querySelectorAll(...)[0] is undefined

This is a situation when frames disapers. For example, an ad on a page served and closed.

Reply:
> > TypeError: document.querySelectorAll(...)[0] is undefined
> 
> This is a situation when frames disapers. For example, an ad on a page served and closed. We should treat it as standard behavior.

For this reason I placed an msgbox 
```autoit
MsgBox($MB_TOPMOST, "", 'Before checking location of multiple elements on multiple frames' & @CRLF & 'Try the same example with and without waiting about 30 seconds in order to see that many frames should be fully loaded, and to check the differences')
```

to give user time manually check and be sure that webpage is completly loaded.

Reply:
have some findigs.

```
__WD_Post: URL=HTTP://127.0.0.1:4444/session/e4b327f4-3c2a-4292-affa-d93f56bac169/execute/sync; Data={"script":"return document.querySelectorAll('iframe')[0].outerHTML;", "args":[]}
__WD_Post ==> Javascript Exception [23] : HTTP status = 500 ResponseText={"value":{"error":"javascript error","message":"TypeError: document.querySelectorAll(...)[0] is undefined","stacktrace":"@https://www.tutorialspoint.com/html/src/creating_frames.htm:2:23\n@https://www.tutorialspoint.com/html/src/creating_frames.htm:3:8\n"}}
_WD_ExecuteScript ==> Javascript Exception [23] : Error occurred when trying to ExecuteScript
- Error occurred on "null/0" level when trying to check attributes child frames #0
```

I add some code to get HTML source when this errors occurs.
And I get:
```html
<html><head>
      <title>HTML Frames</title>
   </head>
	
   <frameset rows="10%,80%,10%">
      <frame name="top" src="/html/top_frame.htm">
      <frame name="main" src="/html/main_frame.htm">
      <frame name="bottom" src="/html/bottom_frame.htm">
   
      <noframes>
         <body>
            Your browser does not support frames.
         </body>
      </noframes>
      
   </frameset>
   
</html>
```

so the problem is that this is `frameset` and we will not be able to get:
`return document.querySelectorAll('iframe')[0].outerHTML;`
as this is not `iframe`

so this is bigger problem...

Reply:
I was trying to use:
`return window.frames[" & $iFrame & "].frameElement.outerHTML;`

but have other things to check.

I wonder if `frameset` and `iframe` can be combined in singe `window.document`

Reply:
I was reading here:
https://stackoverflow.com/a/67765623/5314940
is that mean that `frameset` and `iframe` can not be combined together in single document ?

Reply:
Please take a look here: [<frame..> support](https://github.com/Danp2/au3WebDriver/pull/434/commits/a2d9858e2afd09c0cb16a271d174b2f6d4005687)

Reply:
Untested, but I would think that you could do something like this to eliminate multiple calls to _WD_ExecuteScript --

```autoit
$JS = "let nodes = document.querySelectorAll('iframe'); if (nodes) { return nodes[" & $iFrame & "].outerHTML; } else { return window.frames[" & $iFrame & "].frameElement.outerHTML; }"
$sFrameAttributes = _WD_ExecuteScript($sSession, $JS, Default, Default, $_WD_JSON_Value)
```

Reply:
Thanks. 
Will xheck ASAP

Reply:
Works good.
tested with:
```js
function GetFrame(iFrame) {
  let nodes = document.querySelectorAll('iframe');
  if (nodes.length) { return nodes[iFrame].outerHTML; } 
  else { return window.frames[iFrame].frameElement.outerHTML; }
}

GetFrame(0);
```
Please notice `nodes.length` change.

Reply:
Please take a look on:
[$sJavaScript_OuterHTML](https://github.com/Danp2/au3WebDriver/pull/434/commits/24914dc2bc915e5300d698fc6434975483d7482e)

and also on:
[$sArrayHeader + FRAME](https://github.com/Danp2/au3WebDriver/pull/434/commits/57eba9b91a8813901190ba3f7f1bbf0b5ee864f7)

Reply:
Regarding your recent changes --

- Your usage of `StringFormat` in combination with `_WD_ExecuteScript` looks good.
- Should other coding in the related functions be updated to match this style?
- IMO, when you refer to `frame`, this implies both frames and iframes, so I don't see the need to list both in function header, array header, etc.

Reply:
Please take a look on:
[$sArrayHeader](https://github.com/Danp2/au3WebDriver/pull/434/commits/2b376e02e1cc572187d72f0ee0dca9509bfc40ce)
[header $sFrameAttributes](https://github.com/Danp2/au3WebDriver/pull/434/commits/9b1ef7a24d889ea090737f39fa045c0a1eada5ba)
[$sJavaScript...FrameAttributes...FrameElementID](https://github.com/Danp2/au3WebDriver/pull/434/commits/9ef8c2967a1ab3d69913db685231f998fa51b76f)

Reply:
anything more here ?

Reply:
Not AFAICS. I'm depending on you to perform testing on this since it's your baby.

P.S. Don't forget to remove your debugging code 😉

Reply:
> Not AFAICS. I'm depending on you to perform testing on this since it's your baby.

Just in any case, in future let me know when you giving a green light to clenaup and last review before merge.

> P.S. Don't forget to remove your debugging code 😉


thanks for remider.